### ISSUE TITLE
LargeSample test failes when using shmem and very large samples

### ISSUE THREAD

**Issue by jwillemsen** [objectcomputing/OpenDDS#2799]
When increasing the sample size to around 5Mb with shared memory the Large Sample test fails with the output below, the subscriber doesn't get all samples but doesn't log any errors. Increasing the shared memory pool_size will make the test work again, but I can imagine that it is hard to configure a concrete value. 

We aren't certain whether this failure is correct behavior with the current QoS settings, could the publisher recover this error when the subscriber has read the samples? Are the correct callbacks called?

```
2021-07-02 15:35:04.123@LM_ERROR@exceeded max number of segments = 1, base = 155832320, offset = 16781312
2021-07-02 15:35:04.124@LM_ERROR@(16263|16263) ERROR: ShmemSendStrategy for link 0x6589ff0 failed to allocate 6656103 bytes for data
2021-07-02 15:35:04.125@LM_WARNING@(16263|16263) WARNING: Problem detected in send buffer management: send_bytes: Cannot allocate memory.
2021-07-02 15:35:04.130@LM_DEBUG@
(16263|16263) TransportInst::dump() -
   transport_type:               shmem
   name:                         shmem1
   queue_messages_per_pool:      10
   queue_initial_pools:          5
   max_packet_size:              2147481599
   max_samples_per_packet:       10
   optimum_packet_size:          4096
   thread_per_connection:        false
   datalink_release_delay:       10000
   datalink_control_chunks:      32

   pool_size:                    16777216
   datalink_control_size:        8192
```

**Comment by mitza-oci** [comment]
Shared memory is based on a configurable pool size, and (as seen above) logs an error if the pool runs out.

**Comment by jwillemsen** [comment]
Do we have any unit tests to check what happens when we run out of the pool, what when the subscriber keeps reading, does OpenDDS recover?

**Comment by jwillemsen** [comment]
Got some user feedback that they see @Dynamic_Cached_Allocator_With_Overflow::malloc 174cb1a0 1 heap allocs with 1 outstanding@ in their log, but they are using shared memory transport, wouldn't an allocation from the heap at that moment break shmem?

**Comment by jwillemsen** [comment]
From the same user, when they use a lot of messages it looks the control area runs out of space, that errors looks not to be recovered by OpenDDS

**Comment by jwillemsen** [comment]
Yes, increasing the datalink_control_size did resolve the issue, but with the default size after some time all communication stops, the subscriber doesn't get anything anymore
